### PR TITLE
Add clean option for remove volumes after VM shutdown

### PR DESCRIPTION
**Pull request by sl84702**
It's will save space on the hypervisor.

### CODE REVIEW

**Comment by sl84702** [comment]
And сleaning cache will save space on the client too.

File: run [hunk 1]
```diff
@@ -595,7 +611,10 @@ for dir in "${DIRS[@]}"; do
     remote_path=$(rpath "$dir")
     dname=$(tr / - <<< "${local_path#/}")
     drive="$USR_CACHE_DIR/$VMNAME/$dname.qcow2"
-    drive_uuid=$(drive_uuid "$USR_CACHE_DIR/$VMNAME/$dname.uuid")
+    CACHED+=("$drive")
```
Comment (roolebo) [inline run]: Aren't these going to be deleted by ```vm::vol::clean```?

File: run [hunk 2]
```diff
@@ -584,6 +599,7 @@ run_mkinitrd ${MKINITRD_ARGS[@]} "$INITRAMFS" "$KERNEL"
 VOLUMES+=("$INITRAMFS_VOLUME:$INITRAMFS:raw:")
 
 CRASHDRIVE="$USR_CACHE_DIR/$VMNAME/crash.qcow2"
+CACHED+=("$CRASHDRIVE")
```
Comment (roolebo) [inline run]: I think this one shouldn't be added to the list of cache-related files. It's going to be deleted by ```vm::vol::clean```, right?

**Review by roolebo — CHANGES_REQUESTED** [review]

File: run (hunk 1 above)
Comment (sl84702) [reply run]: These files are on the client.
`vm::vol::clean` delete only volumes on hypervisor and `cache::file::clean` delete only cached files on client.

File: run (hunk 2 above)
Comment (sl84702) [reply run]: No, `vm::vol::clean` delete only volumes from ktest pool on hypervisor

**Review by roolebo — CHANGES_REQUESTED** [review]
There are a few issues left over:
* intird is not getting removed as well as kdump kernel's temporary directory:
```
➜  ktest git:(add-clean) ls -l /tmp/ktest.cache/r.bolshakov/ktest-r.bolshakov
total 37312
drwxr-xr-x  4 r.bolshakov  wheel       128 Feb 14 17:17 boot
-rw-r--r--  1 r.bolshakov  wheel       103 Feb 14 17:17 crash.qcow2.args
-rw-r--r--  1 r.bolshakov  wheel       103 Feb 14 17:17 crash.qcow2.args.tmp
-rw-r--r--  1 r.bolshakov  wheel        37 Feb 14 17:17 crash.uuid
-rw-------@ 1 r.bolshakov  wheel  19050928 Feb 14 17:17 initrd
-rw-r--r--  1 r.bolshakov  wheel       128 Feb 14 17:17 mkinitrd.args
-rw-r--r--  1 r.bolshakov  wheel       128 Feb 14 17:17 mkinitrd.args.tmp
-rw-r--r--  1 r.bolshakov  wheel         0 Feb 14 17:17 port.list
-rw-r--r--  1 r.bolshakov  wheel         6 Feb 14 17:17 port.list--
-rw-r--r--  1 r.bolshakov  wheel         6 Feb 14 17:17 port.list.sorted
-rw-r--r--  1 r.bolshakov  wheel      6000 Feb 14 17:17 port.range
-rw-r--r--  1 r.bolshakov  wheel       387 Feb 14 17:17 qemu-img.log
-rw-r--r--  1 r.bolshakov  wheel         0 Feb 14 17:17 start.log
-rw-r--r--  1 r.bolshakov  wheel      6137 Feb 14 17:17 virsh.log
➜  ktest git:(add-clean) ls -l /tmp/ktest.cache/r.bolshakov/ktest-r.bolshakov/boot
total 22976
-rw-------  1 r.bolshakov  wheel  5548180 Feb 14 17:17 initrd-kdump
-rw-r--r--  1 r.bolshakov  wheel  6213304 Feb 14 17:17 vmlinuz
```
* volumes cache info is stale and references some files that are intended for removal:
```
cat /tmp/ktest.cache/r.bolshakov/volumes
SPB-NB-133.local default kernel-r.bolshakov ad9fd7d052207c18c04be64662f080b2efe35d85
SPB-NB-133.local default initramfs-r.bolshakov 3f1aaede84f85639e659f43309fdd7cebf6facba
SPB-NB-133.local default crashes-r.bolshakov bc5696c8ba47fb8bef285d9c4683ce262a1f9143
```
* the second commit breaks subsequent invocations of ktest:
```
./run --docker-image ktest-opensuse-42.3 -I tests/guest:/test -e /test/selfcheck --clean tests/kernel
Create volume crashes-r.bolshakov on default./run: line 442: /tmp/ktest.cache/r.bolshakov/ktest-r.bolshakov/crash.qcow2: No such file or directory

virsh exited with 1, last log lines:
Type:           file
Capacity:       18.17 MiB
Allocation:     18.17 MiB

2019-02-14T17:27:39Z virsh vol-info crashes-r.bolshakov --pool default
error: failed to get vol 'crashes-r.bolshakov'
error: Storage volume not found: no storage vol with matching path 'crashes-r.bolshakov'

2019-02-14T17:27:39Z virsh vol-create-as --format qcow2 default crashes-r.bolshakov
error: command 'vol-create-as' requires <capacity> option
```

**Comment by sl84702** [comment]
> * intird is not getting removed as well as kdump kernel's temporary directory:

This is done on purpose, as these files do not take up much space.
I think that we should take out their removal somewhere separately.

>* volumes cache info is stale and references some files that are intended for removal:

I agree, I fix it.

>* the second commit breaks subsequent invocations of ktest:
Sorry for that, I tested it only with a --force option.

And what do you think - mb delete all cache (`rm -rf $USR_CACHE_DIR/$VMNAME`)?

**Comment by roolebo** [comment]
> And what do you think - mb delete all cache (`rm -rf $USR_CACHE_DIR/$VMNAME`)?

I'm okay with that. It's going to be possible to debug it in CI environment if clean is not specified.

**Comment by sl84702** [comment]
> > And what do you think - mb delete all cache (`rm -rf $USR_CACHE_DIR/$VMNAME`)?
> 
> I'm okay with that. It's going to be possible to debug it in CI environment if clean is not specified.

Done

File: run [hunk 3]
```diff
@@ -766,6 +773,7 @@ if [ -s $outfile ] && tar -tf $outfile &>/dev/null; then
     if [ -r "$OUTPUT/exit-code" ]; then
         exit $(<"$OUTPUT/exit-code")
     else
+        clean
```
Comment (roolebo) [inline run]: I wonder if we need it. Would the trap above perform the clean up?

**Review by roolebo — CHANGES_REQUESTED** [review]

File: lib [hunk 4]
```diff
@@ -22,6 +22,7 @@ fi
 
 die() {
     echo "$1" >&2
+    clean
```
Comment (roolebo) [inline lib]: I think lib shouldn't  have a dependency on a function that is not defined in the file. That break the whole idea behind a generic library that can be sourced.

**Review by roolebo — CHANGES_REQUESTED** [review]

File: run (hunk 3 above)
Comment (sl84702) [reply run]: Of course:)

**Review by roolebo — APPROVED** [review]
LGTM!

Thank you

**Pull request merged**